cirq/ops/linear_combinations.py

```diff
@@ -111,3 +111,114 @@ def _pauli_expansion_(self) -> value.LinearDict[str]:
+
+    If G1, ..., Gn are gate operations, {q1_1, ..., q1_k1}, {q2_1, ..., q2_k2},
+    ..., {qn_1, ..., qn_kn} are (not necessarily disjoint) sets of qubits and
+    b1, b2, ..., bn are complex numbers, then
+
+        LinearCombinationOfOperations({
+            G1(q1_1, ..., q1_k1): b1,
+            G2(q2_1, ..., q2_k2): b2,
+            ...,
+            Gn(qn_1, ..., qn_kn): bn})
+
+    represents the linear operator
+
+        A = b1 G1(q1_1, ..., q1_k1) +
+          + b2 G2(q2_1, ..., q2_k2) +
+          + ... +
+          + bn Gn(qn_1, ..., qn_kn)
+
+    where in each term qubits not explicitly listed are assumed to be acted on
+    by the identity operator. Note that A may not be unitary or even normal.
+    """
+
+    def __init__(self,
+                 terms: Mapping[raw_types.Operation, value.Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of gate operations to coefficients in the linear
+                combination being initialized.
+        """
+        super().__init__(terms, validator=self._is_compatible)
+
+    def _is_compatible(self, gate: raw_types.Operation) -> bool:
+        return isinstance(gate, raw_types.Operation)
+
+    @property
+    def qubits(self) -> Tuple[raw_types.Qid, ...]:
+        """Returns qubits acted on self."""
+        if not self:
+            return ()
+        qubit_sets = [set(op.qubits) for op in self.keys()]
+        all_qubits = set.union(*qubit_sets)
+        return tuple(sorted(all_qubits))
+
+    def matrix(self) -> np.ndarray:
+        """Reconstructs matrix of self using unitaries of underlying operations.
+
+        Raises:
+            TypeError: if any of the gates in self does not provide a unitary.
+        """
+
+        def extend(u: np.ndarray, qubits: Tuple[raw_types.Qid, ...],
+                   all_qubits: Tuple[raw_types.Qid, ...]) -> np.ndarray:
+            """Extends unitary on qubits to unitary on all_qubits."""
+            assert qubits
+            assert set(qubits).issubset(all_qubits)
+            assert u.shape == (2**len(qubits), 2**len(qubits))
+
+            if qubits[0] != all_qubits[0]:
+                return np.kron(np.eye(2), extend(u, qubits, all_qubits[1:]))
+
+            if len(qubits) == 1:
+                return np.kron(u, np.eye(2**(len(all_qubits) - 1)))
+
+            ab, cd = np.split(u, 2, axis=0)
+            a, b = np.split(ab, 2, axis=1)
+            c, d = np.split(cd, 2, axis=1)
+            a = extend(a, qubits[1:], all_qubits[1:])
+            b = extend(b, qubits[1:], all_qubits[1:])
+            c = extend(c, qubits[1:], all_qubits[1:])
+            d = extend(d, qubits[1:], all_qubits[1:])
+            return np.block([[a, b], [c, d]])
+
+        num_dim = 2**len(self.qubits)
+        result = np.zeros((num_dim, num_dim), dtype=np.complex128)
+        for op, coefficient in self.items():
+            u = protocols.unitary(op)
```

Comment:
This could be made much more efficient by using `apply_unitary`. It would also handle a lot of the expansion logic that you have had to duplicate.

```python
raw = np.eye(num_dim, dtype=np.complex128)
workspace = np.empty_like(raw)
axes = [q_to_i_map[q] for q in op.qubits]
result += protocols.apply_unitary(...Args(raw, workspace, axes))
```

Reply:
Done.